### PR TITLE
github: move static checks and spread over

### PR DESCRIPTION
**Pull request by zyga**
The chain of actions is now:

 - Static checks
 - Unit tests
 - Spread canary (ubuntu 16.04 and core 16)
 - Spread stable (all but canary and unstable)
 - Spread unstable (unstable systems only)

I also switched to ./run-checks --static and --unit to avoid
any drift. I will explore making that faster later on.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -34,10 +33,6 @@ jobs:
       run: sudo apt build-dep -y ${{ github.workspace }}/src/github.com/snapcore/snapd
     - name: Make /var/cache/apt owned by current user
       run: sudo chown -R $(id -u) /var/cache/apt
-    - name: Set up go 1.9
```
Comment (mvo5) [inline .github/workflows/test.yaml:37]: Why is this removed? Is it no longer needed?

Comment (zyga) [reply .github/workflows/test.yaml:37]: Because we get both debian and custom go and I got an error where go said object files were built with different compiler.

Comment (mvo5) [reply .github/workflows/test.yaml:37]: So we rely on running the unit tests in the spread runs?

Comment (zyga) [reply .github/workflows/test.yaml:37]: No, we run unit tests here, look below.

Comment (mvo5) [reply .github/workflows/test.yaml:37]: Sorry if I sound dumb - but how do we know that our stuff still works with go-1.9 if we don't ask for 1.9 in the unit tests check?

Comment (zyga) [reply .github/workflows/test.yaml:37]: We don't, I'm trying to get the best of what's available here.

I'll update this as I'm progressing. I think the tests will actually pass now.

Comment (zyga) [reply .github/workflows/test.yaml:37]: Ok, we discussed this some more and we're now using go 1.10 from the xenial archive. In a follow-up we will switch to go 1.9 from a snap.

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -58,16 +64,83 @@ jobs:
...
+      run: /tmp/spread google:${{ matrix.system }}:tests/...
+    - name: Discard spread workers
+      if: always()
+      run: |
+        shopt -s nullglob;
+        for r in .spread-reuse.*.yaml; do
+          /tmp/spread -discard -reuse-pid="$(echo "$r" | grep -o -E '[0-9]+')";
+        done
+  spread-unstable:
+    needs: [unit-tests, spread-stable]
```
Comment (sergiocazzolato) [inline .github/workflows/test.yaml:141]: make unstable depend on the stable will take longer the whole execution as well

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -58,16 +64,83 @@ jobs:
...
+          cd /tmp
+          curl -s -O https://niemeyer.s3.amazonaws.com/spread-amd64.tar.gz
+          tar xzvf spread-amd64.tar.gz
+          # Register a problem matcher to highlight spread failures
+          echo "::add-matcher::.github/spread-problem-matcher.json"
+    - name: Run spread tests
+      env:
+          SPREAD_GOOGLE_KEY: ${{ secrets.SPREAD_GOOGLE_KEY }}
+      run: /tmp/spread google:${{ matrix.system }}:tests/...
+    - name: Discard spread workers
```
Comment (sergiocazzolato) [inline .github/workflows/test.yaml:133]: I like this

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -58,16 +64,83 @@ jobs:
...
+      run: /tmp/spread -abend google:${{ matrix.system }}:tests/...
+    - name: Discard spread workers
+      if: always()
+      run: |
+        shopt -s nullglob;
+        for r in .spread-reuse.*.yaml; do
+          /tmp/spread -discard -reuse-pid="$(echo "$r" | grep -o -E '[0-9]+')";
+        done
+  spread-stable:
+    needs: [unit-tests, spread-canary]
```
Comment (sergiocazzolato) [inline .github/workflows/test.yaml:101]: make stable depends on the canary will take longer the whole execution, is it any reason for creating this new execution workflow?

**Comment by sergiocazzolato** [comment]
@zyga Thanks for the change, it is really usefull, I left 2 questions inline

**Comment by sergiocazzolato** [comment]
@zyga Thanks for the change, it is really usefull, I left 2 questions inline
